### PR TITLE
Use standard stdlib by default, make libc++ optional

### LINKED ISSUE #86: Support clang with other standard libraries
I've successfully built with clang and have ran the test suite (including TSAN where applicable) on the following platforms:
- Ubuntu 22.04, libstdc++ 11
- Ubuntu 22.04, libstdc++ 12
- Windows, MSVC 14.33.31629*

It seems reasonable to support an OS's defacto stl implementation when possible.

* There is one failure on Windows when using clang with Microsoft's STL arising from the handling of exceptions. Seems like clang doesn't handle things correctly sometimes when exceptions are caught by value. The test suite passes when [this line](https://github.com/David-Haim/concurrencpp/blob/master/test/include/utils/test_ready_lazy_result.h#L63) is changed to catch by const reference rather than value.

### PR DESCRIPTION
**Pull request by chausner**
Closes #86.

Since there is no inbuilt support in CMake to configure the stdlib, the suggested approach is setting the appropriate compile options in a toolchain file.

### CODE REVIEW

File: cmake/coroutineOptions.cmake [hunk 1]
```diff
@@ -5,8 +5,6 @@ function(target_coroutine_options TARGET)
   if(MSVC)
     target_compile_options(${TARGET} PUBLIC /std:c++latest /permissive-)
   elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    target_compile_options(${TARGET} PUBLIC -stdlib=libc++ -fcoroutines-ts)
-    target_link_options(${TARGET} PUBLIC -stdlib=libc++)
```
Comment (chausner) [inline cmake/coroutineOptions.cmake:9]: I think `-fcoroutines-ts` is unnecessary since CMake should add it via `target_compile_features(concurrencpp PUBLIC cxx_std_20)` so I also removed it here.

**Comment by David-Haim** [comment]
I was imagining something a bit different:
on linux, cmake detects the OS and sets libstd++ by default
on macOs, cmake detects the OS and libc++ by default
The user can specify a cmake flag to choose the library .

**Comment by ohanar** [comment]
Another alternative is to just use whatever clang uses by default. Generally that is libstdc++ on Linux, libc++ on MacOs, and msvc's STL on Windows, however, clang can be compiled with a different default stdlib than the platform default.

**Comment by chausner** [comment]
If there is no good reason to use a different runtime than the default, I would also suggest to simply let the compiler pick it by not passing any `--std=` argument at all. This is what my implementation does at the moment. If the user doesn't pass a CMake toolchain file, it will use whatever C++ runtime is the system/compiler default. On macOS, that should be libc++ and on Linux it is usually libstdc++. Since for older operating systems, libstdc++ is too old and does not support coroutines, it can be forced to use libc++ by specifying the toolchain file.

**Comment by David-Haim** [comment]
If I remove `--std=` from the cmake file it fails to compile (it complains it doesn't find `coroutine` and `semaphore` header)

**Comment by chausner** [comment]
> If I remove `--std=` from the cmake file it fails to compile (it complains it doesn't find `coroutine` and `semaphore` header)

Sorry, I meant one should be able to leave out `--stdlib=`. Enabling C++20 features is necessary indeed but in recent CMake versions this is best done via `target_compile_features(concurrencpp PUBLIC cxx_std_20)` instead of `--std=`.

**Comment by David-Haim** [comment]
Sorry for the misunderstanding, I'm really not a cmake expert, but what does that mean in practice? 
how would the cmake/command line look like then?

**Comment by chausner** [comment]
Sorry, it seems we both talked past each other. :) Let me try to explain again.

> I was imagining something a bit different:
on linux, cmake detects the OS and sets libstd++ by default
on macOs, cmake detects the OS and libc++ by default
The user can specify a cmake flag to choose the library .

This is basically how it would work with the changes in this PR. The compiler defaults to libstdc++ on Linux, and to libc++ on macOS. If the user wants to override the default, they can specify a CMake toolchain file via the command-line (`-DCMAKE_TOOLCHAIN_FILE=...`) where any custom compiler settings can be set. We would include a toolchain file for overriding the stdlib to libc++ so that the user does not have to write one by themselves.

This approach is generic and works for any CMake project, as long as the CMake project does not hardcode or override any compiler arguments inside the CMake file (which is what concurrencpp is currently doing).

**Comment by David-Haim** [comment]
@chausner I see.
Can we do better though? Do we really need another file for this to work? Can't we do it with just cmake flags?
Something like `-DUSE_LIBRARY=libc++`?

thanks

**Comment by chausner** [comment]
I think you could override the compiler flags directly on the command-line but I haven't tested it:
```
cmake -DCMAKE_CXX_FLAGS="-stdlib=libc++" -DCMAKE_EXE_LINKER_FLAGS="-stdlib=libc++ -lc++abi"
```

It would also be possible to create a dedicated CMake option for it like `-DUSE_LIBRARY=libc++` but I guess there is a reason why until now CMake still has not added such a standard option. In any case, such options are not standardized. In contrast, `CMAKE_TOOLCHAIN_FILE` is a generic standardized way to configure a compiler toolchain. In our case, we are only setting the stdlib but usually, toolchain files are used to specify many more options like compiler paths, especially for cross-compilation.

If you prefer either of these two alternatives, I can adapt this MR correspondingly.

**Comment by David-Haim** [comment]
OK, sounds legit.
One last question: If I use this library as a git submodule to another project, how do I pass this flag in this case?

**Comment by David-Haim** [comment]
@friendlyanon while you're looking at concurrencpp PRs, mind looking at this one as well?
thanks!

File: README.md [hunk 2]
```diff
@@ -2195,6 +2195,15 @@ $ cd build/test
 $ ctest . -V
 ```
 
+##### Linking to libc++
```
Comment (friendlyanon) [inline README.md:2198]: This doesn't need to be documented in the README. Users who want to use libc++ will have to look up how that can be done with their own compiler and act accordingly and you can't possibly know all the varieties of compilers the users will use. This just adds noise imo.

If I want to use libc++ I will do it my own way, depending on what I'm doing at the time.

File: cmake/coroutineOptions.cmake (hunk 1 above)
Comment (friendlyanon) [reply cmake/coroutineOptions.cmake:9]: Well, it's more like providing flags for specific compilers should be the user's responsibility. Iirc I just copied this bit from what was there previously without major changes.

CMake will not add `-fcoroutines-ts`, the user will have to provide flags according to their own compiler, for which CMake provides many many opportunities.

File: cmake/libc++.cmake [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Specify this file as CMAKE_TOOLCHAIN_FILE when invoking CMake with Clang
```
Comment (friendlyanon) [inline cmake/libc++.cmake:1]: This file should be moved to the `.github` folder, since it's for interal use only in the CI.

**Review by friendlyanon — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (chausner) [reply README.md:2198]: Users will find information how to link to libc++ but it will be much harder to figure out why the library does not build on older libstdc++ versions and how to make it work. I think it would be helpful for a significant number of users if this is explained in the README.

If you prefer to not have explicit instructions how to link to libc++, I'll remove those. But I would keep at least a note that says that it will not work out of the box for these versions.

File: cmake/coroutineOptions.cmake (hunk 1 above)
Comment (chausner) [reply cmake/coroutineOptions.cmake:9]: Yes, I think you are right. I think `-fcoroutines-ts` is only necessary on certain older compiler versions where coroutine support was still experimental. Specifying it for newer versions does not seem to hurt, so we could keep it or we can remove it (with the chance of breaking those older versions).

File: README.md (hunk 2 above)
Comment (friendlyanon) [reply README.md:2198]: Well you bring up a valid point that older libraries will linger around for a while, but personally I would just prefer a notice, instead of explicit instructions.
